### PR TITLE
Changed for loops to iterators

### PR DESCRIPTION
**Pull request by harrydotknee**
Condensed functions in to the format function also.

### CODE REVIEW

File: csv_handler.rb [hunk 1]
```diff
@@ -21,53 +18,44 @@ def loaded?
     @loaded
   end
 
-  # formatting csv file in to 2d array (no titles)
+  # 1) format the csv file in to 2d array (no titles)
   def format
-    output = [@csv_data.length()-1]
-    for i in 1...@csv_data.length()
-      output[i-1] = @csv_data[i]
+    output = []
```
Comment (matt423) [inline csv_handler.rb]: Ruby has `map` which helps us reduce some boilerplate in assigning array outputs, we can use here

File: csv_handler.rb [hunk 2]
```diff
@@ -21,53 +18,44 @@ def loaded?
...
+    @csv_data.drop(1).each do |i|
```
Comment (matt423) [inline csv_handler.rb]: If we specify the CSV as having headers, we could remove this really I guess as we wouldn't need to manually remove the first row

File: csv_handler.rb [hunk 3]
```diff
@@ -21,53 +18,44 @@ def loaded?
...
-  def capitalise_names
-    output = format
-    for i in 0...format.length()
-      # assuming here that we know the name will be in the first column- maybe add checker later
-      # splitting a name and capitalising the first letter of each part of the name
-      split_name = format[i][0].split
+  
+  # 2) capitalise the names
+    
+    output.each do |i|
```
Comment (matt423) [inline csv_handler.rb]: Iterating multiple times over the same object, should be avoided where possible. 
Reasons include performance(if each iteration takes 6 seconds, we'll spend `6 x iteration count` in compute power)

File: csv_handler_spec.rb [hunk 4]
```diff
@@ -3,32 +3,16 @@
...
 describe CsvHandler do
-	subject { CsvHandler.new('data.csv') }
+	subject { CsvHandler.new(file_name) }
+  let(:file_name) { 'data.csv' }
 
   describe '#format' do
-    it 'returns array template' do
-
-      expect(subject.format).to eq([["derek trotter","sa1 1nw", "no"],["Rodney trotter","Sa6 W23","yes"],["Rick Grimes","Sa2 6tr","no"]])    
+    it 'returns contents of csv file in 2d array format' do
```
Comment (matt423) [inline csv_handler_spec.rb]: As the assertion already specifies the result, we can use more human friendly language in context and it descriptions

```suggestion
    it 'formats the CSV correctly' do
```

File: csv_handler_spec.rb [hunk 5]
```diff
@@ -3,32 +3,16 @@
...
+      expect(subject.format).to eq([["Derek Trotter","SA1 1NW", "no"],["Rick Grimes","SA2 6TR","no"]])
```
Comment (matt423) [inline csv_handler_spec.rb]: This is a nice and clear assertion :+1: We could improve formatting a bit by adding new lines between each array

File: csv_handler.rb (hunk 2 above)
Comment (harrydotknee) [reply csv_handler.rb]: trying to implement this, it seems to add more lines of code due to the csv table format

File: csv_handler.rb [hunk 6]
```diff
@@ -2,57 +2,46 @@
...
-  # 1) format the csv file in to 2d array (no titles)
   def format
+  # 1) format the csv file in to 2d array (no titles)
     output = []
-    @csv_data.drop(1).each do |i|
-      output.push(i)
-    end
-
+    @csv_data.map { |row| output.push(row)}
+    output.shift
```
Comment (matt423) [inline csv_handler.rb]:
```suggestion
    @csv_data = @csv_data[1..-1]
```

File: csv_handler.rb [hunk 7]
```diff
@@ -2,72 +2,56 @@
...
   end
 
-  # formatting csv file in to 2d array (no titles)
   def format
-    output = [@csv_data.length()-1]
-    for i in 1...@csv_data.length()
-      output[i-1] = @csv_data[i]
+  # 1) format the csv file in to 2d array (no titles)
+    output = []
+    @csv_data.each do |i|
```
Comment (matt423) [inline csv_handler.rb]: We should still prefer `map` where possible here

File: csv_handler.rb [hunk 8]
```diff
@@ -2,72 +2,56 @@
...
+      row = Hash.new
+      puts i[0]
```
Comment (matt423) [inline csv_handler.rb]: [not blocking] For demo/example tasks debugging output is fine but usually a good habit to remove before committing

File: csv_handler.rb [hunk 9]
```diff
@@ -2,72 +2,56 @@
...
+      row["name"] = i[0]
+      row["postcode"] = i[1]
+      row["opt out"] = i[2]
```
Comment (matt423) [inline csv_handler.rb]: Ruby has another slightly nicer way to construct hashes
```
{
  name: i[0],
  postcode: i[1]
}

```

**Review by matt423 — COMMENTED** [review]
We still iterate over the object multiple times, let's try to condense them further.

Also, let's keep the array format as an option(as well as the hash option)

File: csv_handler.rb (hunk 9 above)
Comment (harrydotknee) [reply csv_handler.rb]: It did not seem to like it when i used this method in paticular, however i used a similair method using '=>'

File: csv_handler.rb [hunk 10]
```diff
@@ -27,4 +28,19 @@ def format
       }
     }[1..-1]
   end
+  
+  def arr_format
+    @csv_data.each do |i|
+      if i[2] == 'yes'
```
Comment (matt423) [inline csv_handler.rb]: The behaviour and logic in here, is duplicated from the hash formatting. 
OOP has a principle, Don't Repeat Yourself(DRY), that is a something we should apply here. Imagine is we wanted another format, we'd re-write functionality for the third time

**Pull request merged**